include/math/SquareRankFourTensor.hpp

```diff
@@ -0,0 +1,271 @@
+     *      - n is compounded in a column major way from k and l, with the restriction k>l
+     */
+    SquareMatrix<double> pairWiseStrictReduce() const {
+
+        // Initialize the resulting matrix
+        const auto K = this->get_dim();
+        SquareMatrix<double> M = SquareMatrix<double>::Zero(K*(K-1)/2, K*(K-1)/2);
+
+        // Calculate the compound indices and bring the elements from the tensor over into the matrix
+        size_t row_index = 0;
+        for (size_t j = 0; j < K; j++) {  // "column major" ordering for row_index<-i,j so we do j first, then i
+            for (size_t i = j+1; i < K; i++) {  // in column major indices, columns are contiguous, so the first of two indices changes more rapidly
+                // require i > j for "lower triangle"
+
+                size_t column_index = 0;
+                for (size_t l = 0; l < K; l++) {  // "column major" ordering for column_index<-k,l so we do l first, then k
+                    for (size_t k = l+1; k < K; k++) {  // in column major indices, columns are contiguous, so the first of two indices changes more rapidly
+                        // require l > k for "lower triangle"
+
+                        M(row_index,column_index) = this->operator()(i,j,k,l);
+
+                        column_index++;
+                    }
+                }
+
+                row_index++;
+            }
+        }
+
+        return M;
+    }
+
+
+    /**
+     *  Add a matrix to a target tensor starting from given indices
+     *
+     *  @tparam r               indicates which starting index (0,1,2,3) should correspond to the 1st matrix index
+     *  @tparam s               indicates which starting index (0,1,2,3) should correspond to the 2nd matrix index
+     *
+     *  @param M                a matrix
+     *  @param i                1st starting index
+     *  @param j                2nd starting index
+     *  @param k                3rd starting index
+     *  @param l                4th starting index
+     *
+     *  @return a reference to updated this
+     *
+     *
+     *  Example:
+     *      Given a rank-4 tensor of dimensions (10,10,10,10), and a matrix of dimensions (3,3)
+     *      When choosing r and s as 2,1 respectively, and i,j,k,l as 1,2,1,0
+     *      The matrix is added to the tensor with 1st index starting at i and 4th index starting at l held fixed at 1 and 0 respectively
+     *      While 2nd index starting from j (2) now is mapped to the 2nd matrix (s=1) element and 3rd index starting from k (1) is mapped to 1st (r=2)
+     *      meaning that the additions follow : tensor(1,2,1,0) += M(0,0), tensor(1,3,1,0) += M(0,1), ..., tensor(1,3,4,0) += M(3,1), tensor(1,4,4,0) += M(3,2), etc
+     *      and for other indices we find : tensor(2,2,1,0) += 0 (no changes, these are obviously not performed)
+     */
+    template<size_t r, size_t s>
+    Self& addBlock(const MatrixX<Scalar>& M, size_t i, size_t j, size_t k, size_t l) {
```

Comment:
This method is not exclusive to a square tensor.

Reply:
Indeed, I missed that one. Done.